File: src/protocol.rs
use std::io::{Read, Write};
use crate::types::Result;

// Simple binary protocol without protobuf
// Message format: [type:u8][len:u32][data]

pub const MSG_META: u8 = 1;
pub const MSG_PREFLIGHT_OK: u8 = 2;
pub const MSG_PREFLIGHT_FAIL: u8 = 3;
pub const MSG_TRANSFER_START: u8 = 4;
pub const MSG_TRANSFER_RESULT: u8 = 5;

#[derive(Debug)]
pub struct FileMeta {
    pub name: String,
    pub size: u64,
    pub is_dir: bool,
}

#[derive(Debug)]
pub struct PreflightOk {
    pub available_space: u64,
}

#[derive(Debug)]
pub struct PreflightFail {
    pub reason: String,
}

#[derive(Debug)]
pub struct TransferStart {
    pub file_size: u64,
}

#[derive(Debug)]
pub struct TransferResult {
    pub ok: bool,
    pub received_bytes: u64,
}

pub fn write_meta<W: Write>(writer: &mut W, meta: &FileMeta) -> Result<()> {
    let name_bytes = meta.name.as_bytes();
    let len = 8 + 1 + 4 + name_bytes.len();
    
    writer.write_all(&[MSG_META])?;
    writer.write_all(&(len as u32).to_be_bytes())?;
    writer.write_all(&meta.size.to_be_bytes())?;
    writer.write_all(&[if meta.is_dir { 1 } else { 0 }])?;
    writer.write_all(&(name_bytes.len() as u32).to_be_bytes())?;
    writer.write_all(name_bytes)?;

    writer.flush()?;
    Ok(())
}

pub fn read_meta<R: Read>(reader: &mut R) -> Result<FileMeta> {
    // Message header (type + length) already read by caller
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    let size = u64::from_be_bytes(buf);
    
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    let is_dir = buf[0] != 0;
    
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    let name_len = u32::from_be_bytes(buf) as usize;
    
    let mut name_buf = vec![0u8; name_len];
    reader.read_exact(&mut name_buf)?;
    let name = String::from_utf8(name_buf).map_err(|_| "Invalid UTF-8")?;
    
    Ok(FileMeta { name, size, is_dir })
}

pub fn write_preflight_ok<W: Write>(writer: &mut W, msg: &PreflightOk) -> Result<()> {
    writer.write_all(&[MSG_PREFLIGHT_OK])?;
    writer.write_all(&8u32.to_be_bytes())?;
    writer.write_all(&msg.available_space.to_be_bytes())?;
    writer.flush()?;
    Ok(())
}

pub fn read_preflight_ok<R: Read>(reader: &mut R) -> Result<PreflightOk> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    let available_space = u64::from_be_bytes(buf);
    Ok(PreflightOk { available_space })
}

pub fn write_preflight_fail<W: Write>(writer: &mut W, msg: &PreflightFail) -> Result<()> {
    let reason_bytes = msg.reason.as_bytes();
    let len = 4 + reason_bytes.len();
    
    writer.write_all(&[MSG_PREFLIGHT_FAIL])?;
    writer.write_all(&(len as u32).to_be_bytes())?;
    writer.write_all(&(reason_bytes.len() as u32).to_be_bytes())?;
    writer.write_all(reason_bytes)?;
    writer.flush()?;
    Ok(())
}

pub fn read_preflight_fail<R: Read>(reader: &mut R) -> Result<PreflightFail> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    let reason_len = u32::from_be_bytes(buf) as usize;
    
    let mut reason_buf = vec![0u8; reason_len];
    reader.read_exact(&mut reason_buf)?;
    let reason = String::from_utf8(reason_buf).map_err(|_| "Invalid UTF-8")?;
    
    Ok(PreflightFail { reason })
}

pub fn write_transfer_start<W: Write>(writer: &mut W, msg: &TransferStart) -> Result<()> {
    writer.write_all(&[MSG_TRANSFER_START])?;
    writer.write_all(&8u32.to_be_bytes())?;
    writer.write_all(&msg.file_size.to_be_bytes())?;
    writer.flush()?;
    Ok(())
}

pub fn read_transfer_start<R: Read>(reader: &mut R) -> Result<TransferStart> {
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    let file_size = u64::from_be_bytes(buf);
    Ok(TransferStart { file_size })
}

pub fn write_transfer_result<W: Write>(writer: &mut W, msg: &TransferResult) -> Result<()> {
    writer.write_all(&[MSG_TRANSFER_RESULT])?;
    writer.write_all(&9u32.to_be_bytes())?;
    writer.write_all(&[if msg.ok { 1 } else { 0 }])?;
    writer.write_all(&msg.received_bytes.to_be_bytes())?;
    writer.flush()?;
    Ok(())
}

pub fn read_transfer_result<R: Read>(reader: &mut R) -> Result<TransferResult> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    let ok = buf[0] != 0;
    
    let mut buf = [0u8; 8];
    reader.read_exact(&mut buf)?;
    let received_bytes = u64::from_be_bytes(buf);
    
    Ok(TransferResult { ok, received_bytes })
}

pub fn read_message_type<R: Read>(reader: &mut R) -> Result<u8> {
    let mut buf = [0u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

pub fn read_message_length<R: Read>(reader: &mut R) -> Result<u32> {
    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(u32::from_be_bytes(buf))
}

pub fn write_raw_bytes<W: Write>(writer: &mut W, data: &[u8]) -> Result<()> {
    writer.write_all(data)?;
    writer.flush()?;
    Ok(())
}

pub fn read_exact_bytes<R: Read>(reader: &mut R, buf: &mut [u8]) -> Result<()> {
    reader.read_exact(buf)?;
    Ok(())
}

File: csrc/protocol.c
#include "protocol.h"
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#ifdef __APPLE__
#include <libkern/OSByteOrder.h>
#define my_htonll(x) OSSwapHostToBigInt64(x)
#define my_ntohll(x) OSSwapBigToHostInt64(x)
#else
#include <arpa/inet.h>  // for htonl, ntohl

// Helper functions for byte order conversion
static uint64_t my_htonll(uint64_t x) {
    if (htonl(1) == 1) return x;  // If we're on a big-endian system
    return ((uint64_t)htonl(x & 0xFFFFFFFF) << 32) | htonl(x >> 32);
}

static uint64_t my_ntohll(uint64_t x) {
    return my_htonll(x);  // Network to host is the same operation as host to network
}
#endif

// Memory management functions
void file_meta_free(FileMeta* meta) {
    if (!meta) return;
    free(meta->name);
    meta->name = NULL;
}

void preflight_fail_free(PreflightFail* msg) {
    if (!msg) return;
    free(msg->reason);
    msg->reason = NULL;
}

// Helper function to write a string with length prefix
static int write_string(FILE* writer, const char* str) {
    if (!str) {
        uint32_t zero = 0;
        if (fwrite(&zero, 4, 1, writer) != 1) return -1;
        return 0;
    }
    
    uint32_t len = strlen(str);
    uint32_t len_be = htonl(len);
    
    if (fwrite(&len_be, 4, 1, writer) != 1) return -1;
    if (len > 0 && fwrite(str, 1, len, writer) != len) return -1;
    
    return 0;
}

// Helper function to read a string with length prefix
static int read_string(FILE* reader, char** str) {
    uint32_t len_be;
    if (fread(&len_be, 4, 1, reader) != 1) return -1;
    
    uint32_t len = ntohl(len_be);
    if (len == 0) {
        *str = NULL;
        return 0;
    }
    
    char* buf = malloc(len + 1);
    if (!buf) return -1;
    
    if (fread(buf, 1, len, reader) != len) {
        free(buf);
        return -1;
    }
    
    buf[len] = '\0';
    *str = buf;
    return 0;
}

int write_meta(FILE* writer, const FileMeta* meta) {
    if (!writer || !meta) {
        errno = EINVAL;
        return -1;
    }
    
    uint32_t name_len = meta->name ? strlen(meta->name) : 0;
    uint32_t len = 8 + 1 + 1 + 4 + name_len;  // size(8) + is_dir(1) + overwrite_mode(1) + name_len(4) + name
    uint32_t len_be = htonl(len);
    
    uint8_t type = MSG_META;
    uint64_t size_be = my_htonll(meta->size);
    uint8_t is_dir = meta->is_dir ? 1 : 0;
    uint8_t overwrite = meta->overwrite_mode;
    
    if (fwrite(&type, 1, 1, writer) != 1) return -1;
    if (fwrite(&len_be, 4, 1, writer) != 1) return -1;
    if (fwrite(&size_be, 8, 1, writer) != 1) return -1;
    if (fwrite(&is_dir, 1, 1, writer) != 1) return -1;
    if (fwrite(&overwrite, 1, 1, writer) != 1) return -1;
    if (write_string(writer, meta->name) != 0) return -1;
    
    fflush(writer);
    return 0;
}

int read_meta(FILE* reader, FileMeta* meta) {
    if (!reader || !meta) {
        errno = EINVAL;
        return -1;
    }
    
    uint64_t size_be;
    uint8_t is_dir;
    uint8_t overwrite;
    
    if (fread(&size_be, 8, 1, reader) != 1) return -1;
    if (fread(&is_dir, 1, 1, reader) != 1) return -1;
    if (fread(&overwrite, 1, 1, reader) != 1) return -1;
    
    meta->size = my_ntohll(size_be);
    meta->is_dir = is_dir != 0;
    meta->overwrite_mode = overwrite;
    
    if (read_string(reader, &meta->name) != 0) return -1;
    
    return 0;
}

int write_preflight_ok(FILE* writer, const PreflightOk* msg) {
    if (!writer || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Writing PreflightOK message (space=%llu)\n", 
            (unsigned long long)msg->available_space);
    uint8_t type = MSG_PREFLIGHT_OK;
    uint32_t len = 8;
    uint32_t len_be = htonl(len);
    uint64_t space_be = my_htonll(msg->available_space);
    
    fprintf(stderr, "DEBUG: Protocol: Writing type=%d, len=%u, space_be=0x%016llx\n", 
            type, len, (unsigned long long)space_be);
    
    if (fwrite(&type, 1, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write type (errno: %d)\n", errno);
        return -1;
    }
    if (fwrite(&len_be, 4, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write length (errno: %d)\n", errno);
        return -1;
    }
    if (fwrite(&space_be, 8, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write space (errno: %d)\n", errno);
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: PreflightOK message written, flushing\n");
    if (fflush(writer) != 0) {
        fprintf(stderr, "DEBUG: Protocol: Failed to flush PreflightOK (errno: %d)\n", errno);
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: PreflightOK flushed successfully\n");
    return 0;
}

int read_preflight_ok(FILE* reader, PreflightOk* msg) {
    if (!reader || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    uint64_t space_be;
    if (fread(&space_be, 8, 1, reader) != 1) return -1;
    
    msg->available_space = my_ntohll(space_be);
    return 0;
}

int write_preflight_fail(FILE* writer, const PreflightFail* msg) {
    if (!writer || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    uint32_t reason_len = msg->reason ? strlen(msg->reason) : 0;
    uint32_t len = 4 + reason_len;
    uint32_t len_be = htonl(len);
    
    uint8_t type = MSG_PREFLIGHT_FAIL;
    
    if (fwrite(&type, 1, 1, writer) != 1) return -1;
    if (fwrite(&len_be, 4, 1, writer) != 1) return -1;
    if (write_string(writer, msg->reason) != 0) return -1;
    
    fflush(writer);
    return 0;
}

int read_preflight_fail(FILE* reader, PreflightFail* msg) {
    if (!reader || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    return read_string(reader, &msg->reason);
}

int write_transfer_start(FILE* writer, const TransferStart* msg) {
    if (!writer || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    uint8_t type = MSG_TRANSFER_START;
    uint32_t len = 8;
    uint32_t len_be = htonl(len);
    uint64_t size_be = my_htonll(msg->file_size);
    
    if (fwrite(&type, 1, 1, writer) != 1) return -1;
    if (fwrite(&len_be, 4, 1, writer) != 1) return -1;
    if (fwrite(&size_be, 8, 1, writer) != 1) return -1;
    
    fflush(writer);
    return 0;
}

int read_transfer_start(FILE* reader, TransferStart* msg) {
    if (!reader || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    uint64_t size_be;
    if (fread(&size_be, 8, 1, reader) != 1) return -1;
    
    msg->file_size = my_ntohll(size_be);
    return 0;
}

int write_transfer_result(FILE* writer, const TransferResult* msg) {
    if (!writer || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Writing TransferResult message (ok=%d, bytes=%llu)\n", 
            msg->ok, (unsigned long long)msg->received_bytes);
    uint8_t type = MSG_TRANSFER_RESULT;
    uint32_t len = 9;
    uint32_t len_be = htonl(len);
    uint8_t ok = msg->ok ? 1 : 0;
    uint64_t bytes_be = my_htonll(msg->received_bytes);
    
    fprintf(stderr, "DEBUG: Protocol: Writing type=%d, len=%u, ok=%d, bytes_be=0x%016llx\n", 
            type, len, ok, (unsigned long long)bytes_be);
    
    if (fwrite(&type, 1, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write type (errno: %d)\n", errno);
        return -1;
    }
    if (fwrite(&len_be, 4, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write length (errno: %d)\n", errno);
        return -1;
    }
    if (fwrite(&ok, 1, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write ok flag (errno: %d)\n", errno);
        return -1;
    }
    if (fwrite(&bytes_be, 8, 1, writer) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to write bytes (errno: %d)\n", errno);
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: TransferResult written, flushing\n");
    if (fflush(writer) != 0) {
        fprintf(stderr, "DEBUG: Protocol: Failed to flush TransferResult (errno: %d)\n", errno);
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: TransferResult flushed successfully\n");
    return 0;
}

// Helper function to read transfer result with full message handling
int read_transfer_result_full(FILE* reader, TransferResult* msg) {
    if (!reader || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Reading TransferResult message (full)\n");
    
    uint8_t msg_type;
    if (read_message_type(reader, &msg_type) != 0) {
        fprintf(stderr, "DEBUG: Protocol: Failed to read message type (errno: %d)\n", errno);
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: Read message type: %d\n", msg_type);
    
    if (msg_type != MSG_TRANSFER_RESULT) {
        fprintf(stderr, "DEBUG: Protocol: Expected MSG_TRANSFER_RESULT (%d), got %d\n", 
                MSG_TRANSFER_RESULT, msg_type);
        errno = EPROTO;
        return -1;
    }
    
    uint32_t msg_len;
    if (read_message_length(reader, &msg_len) != 0) {
        fprintf(stderr, "DEBUG: Protocol: Failed to read message length (errno: %d)\n", errno);
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: Read message length: %u\n", msg_len);
    
    return read_transfer_result_payload(reader, msg);
}

// Helper function to read just the payload part
int read_transfer_result_payload(FILE* reader, TransferResult* msg) {
    if (!reader || !msg) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Reading TransferResult payload\n");
    uint8_t ok;
    uint64_t bytes_be;
    
    // Read payload only - header should already be read by caller
    if (fread(&ok, 1, 1, reader) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to read ok flag (errno: %d, ferror: %d, feof: %d)\n", 
                errno, ferror(reader), feof(reader));
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: Read ok flag: %d\n", ok);
    
    if (fread(&bytes_be, 8, 1, reader) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to read bytes (errno: %d, ferror: %d, feof: %d)\n", 
                errno, ferror(reader), feof(reader));
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: Read bytes value (raw: 0x%016llx)\n", 
            (unsigned long long)bytes_be);
    
    msg->ok = ok != 0;
    msg->received_bytes = my_ntohll(bytes_be);
    fprintf(stderr, "DEBUG: Protocol: TransferResult payload read successfully (ok=%d, bytes=%llu)\n",
            msg->ok, (unsigned long long)msg->received_bytes);
    return 0;
}

int read_transfer_result(FILE* reader, TransferResult* msg) {
    return read_transfer_result_payload(reader, msg);
}

int read_message_type(FILE* reader, uint8_t* type) {
    if (!reader || !type) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Reading message type\n");
    size_t bytes_read = fread(type, 1, 1, reader);
    if (bytes_read != 1) {
        if (feof(reader)) {
            fprintf(stderr, "DEBUG: Protocol: End of stream reached (connection closed)\n");
        } else {
            fprintf(stderr, "DEBUG: Protocol: Failed to read message type (bytes_read=%zu, errno=%d, ferror=%d, feof=%d)\n", 
                    bytes_read, errno, ferror(reader), feof(reader));
        }
        return -1;
    }
    fprintf(stderr, "DEBUG: Protocol: Read message type: %d\n", *type);
    return 0;
}

int read_message_length(FILE* reader, uint32_t* length) {
    if (!reader || !length) {
        errno = EINVAL;
        return -1;
    }
    
    fprintf(stderr, "DEBUG: Protocol: Reading message length\n");
    uint32_t len_be;
    if (fread(&len_be, 4, 1, reader) != 1) {
        fprintf(stderr, "DEBUG: Protocol: Failed to read message length (errno=%d, ferror=%d, feof=%d)\n", 
                errno, ferror(reader), feof(reader));
        return -1;
    }
    
    *length = ntohl(len_be);
    fprintf(stderr, "DEBUG: Protocol: Read message length: %u (raw: 0x%08x)\n", *length, len_be);
    return 0;
}

int write_raw_bytes(FILE* writer, const uint8_t* data, size_t len) {
    if (!writer || (!data && len > 0)) {
        errno = EINVAL;
        return -1;
    }
    
    if (len > 0 && fwrite(data, 1, len, writer) != len) return -1;
    fflush(writer);
    return 0;
}

int read_exact_bytes(FILE* reader, uint8_t* buf, size_t len) {
    if (!reader || (!buf && len > 0)) {
        errno = EINVAL;
        return -1;
    }
    
    return len > 0 && fread(buf, 1, len, reader) != len ? -1 : 0;
}
